### PR TITLE
Oxlint JS Plugins Alpha

### PR DESCRIPTION
**Pull request by Boshen**
Blog post for Oxlint JS plugins alpha release.

### CODE REVIEW

**Comment by overlookmotel** [comment]
I've pushed a commit with a **very** rough draft. Lots of gaps, and it's not slick writing, lots of repetition.

Purpose is to communicate the broad strokes of what I think the thrust should be.

In short: We need to set expectations about performance - this release is about API coverage and JS plugins actually usable in the real world, not stellar performance.

That said, we've not done a lot of benchmarking yet. Once we do, if it performs better/worse than expected, we can tilt the narrative a bit.

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+Last year, we released the first technical preview of our solution - support for Oxlint plugins written in Javascript, _and_ compatible with ESLint's plugin API.
+
+However, that initial preview was incomplete. Many APIs were not implemented yet.
+
+Since then, we've been working hard on filling out the whole API surface, ramping up performance, and thoroughly testing the implementation with popular ESLint plugins. At this point, we feel it's ready to try out in real world projects.
+
+We expect 80% of users will find they are now able to switch from ESLint to Oxlint and it should "just work".
+
+Oxlint supports over 600 popular rules re-implemented in Rust, which run at native speed. JS plugins aim to "fill in the gaps" where Oxlint does not yet support all the rules users need. The combination of raw native performance for the majority of lint rules, and the flexibility of JS plugins for the rest, aims to make Oxlint "the best of both worlds".
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: If this isn't being published for another week and a half, I think you could reasonably say 700? (though I guess 20ish are oxc rules, not ports)

```suggestion
Oxlint supports over 650 popular rules re-implemented in Rust, which run at native speed. JS plugins aim to "fill in the gaps" where Oxlint does not yet support all the rules users need. The combination of raw native performance for the majority of lint rules, and the flexibility of JS plugins for the rest, aims to make Oxlint "the best of both worlds".
```

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+Many projects have seen significant performance improvements switching from ESLint to Oxlint. See [performance](#performance) section below.
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]:
```suggestion
Many projects have seen significant performance improvements switching from ESLint to Oxlint. See the [performance](#performance) section below.
```

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+We strongly recommend moving from using the linter for code formatting, to using a dedicated formatter, if you can.
+
+A dedicated formatter written in a native language will be an order of magnitude faster than linter plugins like ESLint Stylistic.
+
+Obviously, we would recommend [Oxfmt](../docs/guide/usage/formatter)! Oxlint and Oxfmt make a very strong team.
+
+### Perf tip 2: Choose plugins wisely
+
+Contrary to what many believe, it is perfectly possible to write extremely performant Javascript code. Oxlint is not fast just because it's written in Rust, it's also carefully designed with performance in mind.
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]:
```suggestion
Contrary to what many believe, it is perfectly possible to write extremely performant JavaScript code. Oxlint is not fast just because it's written in Rust, it's also carefully designed with performance in mind.
```

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+The code of JS plugins you select to use in your project is not under Oxlint's control, and to get good performance out of Oxlint overall requires the JS code you ask Oxlint to run to perform well too.
+
+If a plugin uses inefficient algorithms or, for example, performs a lot of filesystem operations, it'll likely be slow in ESLint, and it will be slow in Oxlint too. What Oxlint _can_ do is provide a lightning-fast parser and performant APIs for plugins to interface with, but it can't magically make slow JS code faster.
+
+We will in future provide a utility to diagnose which plugins/rules are the performance bottlenecks in your project, if you find linting is not as fast as you'd like.
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]:
```suggestion
In the future, we will provide a utility to diagnose which plugins/rules are the performance bottlenecks in your project, if you find that linting is not running as fast as you'd like.
```

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+Or manually create a config file:
+
+```json [.oxlintrc.json]
+{
+  "jsPlugins": ["eslint-plugin-react-hooks"],
+  "rules": {
+    "react-hooks/error-boundaries": "error"
+  }
+}
+```
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: I thought this would fail due to `react-hooks` being a native scope?

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+git checkout bench-oxlint
+hyperfine -i --warmup 1 --runs 5 "node --run oxlint"
+```
+
+:::
+
+</details>
+</div>
+
+Projects which currently use TypeScript-ESLint will likely see much larger performance gains. Gains of up to 100x have been reported.
```
Comment (connorshea) [inline src/blog/2026-03-20-oxlint-js-plugins-alpha.md]:
```suggestion
Projects which currently use TypeScript-ESLint or eslint-plugin-import will likely see much larger performance gains. Gains of up to 100x have been reported.
```

**Review by connorshea — COMMENTED** [review]
Exciting to see this come together!! Left some suggestions and nits :)

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+- [Testing Library](https://www.npmjs.com/package/eslint-plugin-testing-library): 17,000 tests, 100% pass.
+- [SonarJS](https://www.npmjs.com/package/eslint-plugin-sonarjs): 4,000 tests, 99.6% pass (excluding type-aware rules).
+- [e18e](https://www.npmjs.com/package/@e18e/eslint-plugin): 500 tests, 100% pass (excluding type-aware rules).
+
+Just because a plugin isn't included in the list above, doesn't mean it doesn't work. Most likely we just haven't tested it yet. ESLint's own tests cover the entire API surface, so a 100% pass rate gives us confidence that we've covered corner cases, as well as the happy path. Please try it out and let us know!
+
+### What it can't do (yet)
+
+- Limited support for front-end frameworks' custom file formats (e.g. Svelte, Vue, Angular) - coming later this year.
+- No custom type-aware rules (TypeScript ESLint's rules are supported via [type-aware linting](../docs/guide/usage/linter/type-aware)).
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: 59/61 is correct, but I think we can elide that detail here for brevity. "already built into Oxlint" is better though - have amended.

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+## Performance
+
+Since the first technical preview, we have "Rustified" large chunks of the code powering JS plugins, which has delivered significant performance gains. In particular, plugins which rely on tokens APIs (e.g. ESLint Stylistic) are up to 5 times faster than before.
+
+As a benchmark, we migrated Node.js's repo from ESLint to Oxlint. Node.js is a large project utilizing many custom lint rules, as well as several large ESLint plugins.
+
+| Linter | Time                 | Speed-up |
+| ------ | -------------------- | -------- |
+| ESLint | 1 minute, 43 seconds |          |
+| Oxlint | 21 seconds           | 4.8x     |
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: Good point! I'll get the stats on that and add them.

[the remark this answers is not in the document]
Comment (overlookmotel) [reply src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: Added. 98 JS rules in total.

File: src/blog/2026-03-20-oxlint-js-plugins-alpha.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+Oxlint JS plugins support is tested against the full test suite of ESLint itself, and also against a wide selection of ESLint plugins, including:
+
+- ESLint built-in rules: 33,000 tests, 100% pass.
+- [React hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks) (including React Compiler rules): 5,000 tests, 100% pass.
+- [ESLint Stylistic](https://eslint.style/): 18,000 tests, 99.99% pass.
+- [Testing Library](https://www.npmjs.com/package/eslint-plugin-testing-library): 17,000 tests, 100% pass.
+- [SonarJS](https://www.npmjs.com/package/eslint-plugin-sonarjs): 4,000 tests, 99.6% pass (excluding type-aware rules).
+- [e18e](https://www.npmjs.com/package/@e18e/eslint-plugin): 500 tests, 100% pass (excluding type-aware rules).
+
+Just because a plugin isn't included in the list above, doesn't mean it doesn't work. Most likely we just haven't tested it yet. ESLint's own tests cover the entire API surface, so a 100% pass rate gives us confidence that we've covered corner cases, as well as the happy path. Please try it out and let us know!
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply src/blog/2026-03-20-oxlint-js-plugins-alpha.md]: We've decided to close https://github.com/oxc-project/oxc/discussions/14862.
